### DISCUSSION TITLE
Getting time offsets of beginning and end of each word

### DISCUSSION THREAD

**Discussion by shashanoid** [openai/whisper#3] [Q&A] [answered]
Hello, I was wondering if it would be possible to get time offsets of start & end of each word / sentences as they appear in the audio.

# Motivation

I was exploring google's https://cloud.google.com/speech-to-text/docs/async-time-offsets and thought it would be great if whisper can produce a similar dataset.

**Comment by aliJabra** [comment]
Looks like it’s already supported. See the librispeech notebook (or colab example), it’s passed as:

options = whisper.DecodingOptions(language=\"en\", without_timestamps=True)

Setting the flag False should return what you want.

**Reply by R4ZZ3** [reply]
Great! This was one question I also had

**Reply by R4ZZ3** [reply]
![image](https://user-images.githubusercontent.com/25264037/191590816-512f4143-fbc9-46af-b32b-6e86a619e760.png)
But actually it seems it is just using those to boost performance somehow but not actually returning the timestamps. Or at least I could not figure out how to get those

**Reply by jongwook** [reply]
Could you check if `transcription["segments"]` has the timestamps? Currently the implementation produces phrase-level timestamps.

**Reply by aliJabra** [reply]
Yes, in transcribe.py:

Returns
-------
A dictionary containing the resulting text ("text") and segment-level details ("segments"), and
the spoken language ("language"), which is detected when `decode_options["language"]` is None.
"""

On a side note, how is the performance in Finnish? Looks like transcription is pretty close from your screenshot.

**Reply by R4ZZ3** [reply]
Thanks @jongwook,
I was able to retrieve segments.
However, one note. My audio is only 9 seconds long. I think it would I quess make more sence to calculate the duration of the last segment without padding?
![image](https://user-images.githubusercontent.com/25264037/191598496-fcb1d6e1-5bd0-49c1-9181-4808f0db1068.png)


@aliJabra 
The one below is not actual transcription. 
It is the output with without_timestamps=False. But so far seems somewhat promising, especially if I can finetune this more and then run some benchmarks against our wav2vec2 models (Some common voice benchmarks here ) https://huggingface.co/Finnish-NLP/wav2vec2-base-fi-voxpopuli-v2-finetuned

**Comment by jongwook** [comment]
We are sampling timestamp tokens mixed with text tokens, which provides phrase-level timestamps like:

```
...
[08:23.000 --> 08:26.000]  I mean, you know, it will break you down.
[08:26.000 --> 08:29.000]  Beat you up and smash you to the floor.
[08:29.000 --> 08:32.000]  You know, but you just got to persevere and fight through it.
[08:32.000 --> 08:35.000]  And I was close many times to say, why am I doing this?
[08:35.000 --> 08:37.000]  It makes no sense, you know.
[08:37.000 --> 08:41.000]  I guess if you can do it, though, and it's available to you,
[08:41.000 --> 08:42.000]  you're going to, right?
[08:42.000 --> 08:43.000]  Yeah.
...
```

Getting word-level timestamps are not directly supported, but it could be possible using the predicted distribution over the timestamp tokens or the cross-attention weights.

**Reply by usergit** [reply]
this would be the single biggest addition to this repo, it could be transformative for audio editing, podcasting etc... I know It's a big ask, but if you could provide a code example, it doesn't even have to be working, it could even be a pseudo code, It would be great.

**Reply by lectair** [reply]
I've written a small script that converts the output to an SRT file. It is useful for getting subtitles in a universal format for any audio:
```
from datetime import timedelta
import os
import whisper

def transcribe_audio(path):
    model = whisper.load_model("base") # Change this to your desired model
    print("Whisper model loaded.")
    transcribe = model.transcribe(audio=path)
    segments = transcribe['segments']

    for segment in segments:
        startTime = str(0)+str(timedelta(seconds=int(segment['start'])))+',000'
        endTime = str(0)+str(timedelta(seconds=int(segment['end'])))+',000'
        text = segment['text']
        segmentId = segment['id']+1
        segment = f"{segmentId}\n{startTime} --> {endTime}\n{text[1:] if text[0] is ' ' else text}\n\n"

        srtFilename = os.path.join("SrtFiles", f"VIDEO_FILENAME.srt")
        with open(srtFilename, 'a', encoding='utf-8') as srtFile:
            srtFile.write(segment)

    return srtFilename
```
Sample SRT File:
```
1
00:00:00,000 --> 00:00:05,000
Open AI has recently decided to open source.

2
00:00:05,000 --> 00:00:09,000
Their translation and transcription AI whisper.

3
00:00:09,000 --> 00:00:18,000
So now it is under an MIT license and that includes both the code that's here as well as the model weights that were used to train the AI.

4
00:00:18,000 --> 00:00:26,000
So if you want it to go and try and make your own speech transcription AI with that data, you are free to do so.
```

**Reply by RaulKite** [reply]
That's not the same. 

It will really useful to have the start and end of each word not sentences.

**Comment by johnafish** [comment]
I hacked together a script today using whisper with Wav2Vec2 forced alignment (https://pytorch.org/tutorials/intermediate/forced_alignment_with_torchaudio_tutorial.html) that generates word-level srt captions. Feel free to play with it and modify it, I might spend a couple more hours cleaning it up and making it more robust but I'm leaving it here for now: https://github.com/johnafish/whisperer

**Comment by jianfch** [comment, marked as answer]
**Update with full script**
https://github.com/jianfch/stable-ts

You can actually get the timestamp prediction for each word because it's part of the predictions but it's filtered and reserved for the start time and end time tokens. That means you can clone the logits to filter it then return it along with the other results.

Add those lines marked with  "# <----add this" in decoding.DecodingTask._main_loop:
```
    def _main_loop(self, audio_features: Tensor, tokens: Tensor):
        assert audio_features.shape[0] == tokens.shape[0]
        n_batch = tokens.shape[0]
        sum_logprobs: Tensor = torch.zeros(n_batch, device=audio_features.device)
        no_caption_probs = [np.nan] * n_batch

        ts_tokens = None  # <----add this
        try:
            for i in range(self.sample_len):
                logits = self.inference.logits(tokens, audio_features)

                if i == 0 and self.tokenizer.no_captions is not None:  # save no_caption_probs
                    probs_at_sot = logits[:, self.sot_index].float().softmax(dim=-1)
                    no_caption_probs = probs_at_sot[:, self.tokenizer.no_captions].tolist()

                # now we need to consider the logits at the last token only
                logits = logits[:, -1]

                logits_clone = torch.clone(logits)  # <----add this
                for k in range(tokens.shape[0]):  # <----add this
                    logits_clone[k, : self.tokenizer.timestamp_begin] = -np.inf  # <----add this
                ts_token = torch.argmax(logits_clone,  dim=-1)[:, None]  # <----add this
                ts_tokens = ts_token if ts_tokens is None else torch.cat([ts_tokens, ts_token], dim=-1)  # <----add this
                del logits_clone  # <----add this

                # apply the logit filters, e.g. for suppressing or applying penalty to
                for logit_filter in self.logit_filters:
                    logit_filter.apply(logits, tokens)

                # expand the tokens tensor with the selected next tokens
                tokens, completed = self.decoder.update(tokens, logits, sum_logprobs)

                if completed or tokens.shape[-1] > self.n_ctx:
                    break
        finally:
            self.inference.cleanup_caching()

        return tokens, sum_logprobs, no_caption_probs, ts_tokens  # <----add this
```
there a couple more methods you'll need to return through to get back up to transcribe()
which is where the calculation for the timestamps occur
time_of_a_word = timestamp_offset + (timestamp_token -  tokenizer.timestamp_begin) * time_precision

~~*note: i haven't full test this yet. just wrote this when i was debugging. but i will do this when i have more time~~

**Reply by lectair** [reply]
Very useful, thanks

**Reply by Saccarab** [reply]
ts_token = torch.argmax(logits_clone)[:, None]  # <----add this

I'll crash with "too many indices for tensor of dimension 0" on the line above.

**Reply by RaulKite** [reply]
> Very useful, thanks

Will you make a diff of this patch working? Please 

Thanks

**Reply by jianfch** [reply]
> ts_token = torch.argmax(logits_clone)[:, None] # <----add this
> 
> I'll crash with "too many indices for tensor of dimension 0" on the line above.

change the line to:
```
logits_clone = torch.clone(logits[:, -1])  # <----add this
```
or move that whole chunk of added code to after the line:
```
logits = logits[:, -1]
```

also change 0 to k the following line:
```
logits_clone[k, : self.tokenizer.timestamp_begin] = -np.inf
```

i notice one of the added lines is also missing dim=-1 argument
```
ts_token = torch.argmax(logits_clone, dim=-1)[:, None]  # <----add this
```

**Reply by ClashSAN** [reply]
Could these be time released in chunks like youtube auto-captions, rather than RSVP (rapid serial visual presentation)? 
This will help for speed reading but you still have buffer room to look at the previous text.

**Reply by jianfch** [reply]
so it seems like from limited testing i'v done, the model size crucial to getting accurate the word timestamps
Large model
![image](https://user-images.githubusercontent.com/28970749/192404569-1c951ee3-8ad6-414c-8cdc-7f5773075bbc.png)
Base model _(not sure whats causing the second portion to go over audio duration)_
![image](https://user-images.githubusercontent.com/28970749/192404674-d0443b5e-1dd8-4af9-b303-bd99b4e9de5b.png)

**Reply by RaulKite** [reply]
Can you provide a diff of the source code apply it? 

Thanks

**Reply by jianfch** [reply]
> Can you provide a diff of the source code apply it?
> 
> Thanks

i managed to stabilize the wobbly timestamps with multiple predictions for each word. just sorting out a few bugs. i'll probably upload the code some time tmr.

**Reply by FurkanGozukara** [reply]
I am looking forward to as well. I think Whisper should an option to output timestamps instead of sentences

**Reply by jianfch** [reply]
there might still be a few bugs, so let me know if stuff breaks and seems off 
https://github.com/jianfch/stable-ts

**Reply by eschmidbauer** [reply]
![image](https://user-images.githubusercontent.com/7139998/193068169-53e43a47-3743-4d65-a838-fcc37b0f5ba3.png)
Hi @jianfch , im testing your code and i get a lot of words with same start timestamp. Any ideas on that? 
using `large` model
calling `transcribe` with defaults like
```
        result = model.transcribe(
            input, verbose=True)
```

**Reply by jianfch** [reply]
@eschmidbauer 
you can check the original results and compare to see what's going on
```python
print('\n'.join([[f'{ts["word"]}: {ts["timestamps"][0]}' for ts in seg['unstable_word_timestamps']] for seg in result['segments'] if seg['end'] > 46 > seg['start']][0]))
```

very likely those are the original results but with the incoherent timestamps replaced or else it might be that of those timestamps went over the 49.02sec mark so they got max clamped to 49.02

**Reply by ArtyomZemlyak** [reply]
I think its bc in unstable_word_timestamps a lot of words just dont have good ts. In this cases taked ts from other words.

**Reply by eschmidbauer** [reply]
![image](https://user-images.githubusercontent.com/7139998/193086084-13da2fd3-313f-4db2-9164-4790ffeff0b2.png)
![image](https://user-images.githubusercontent.com/7139998/193086123-6603f53a-1e2a-4b96-bb80-cbe1176d4ce4.png)

interesting...1st screenshot is the verbose output, the 2nd screenshot is output from your `print()` above

jumps from 48 -> 67

**Reply by ArtyomZemlyak** [reply]
Additional. In my case on another language words more like a tokens. Something like this: "bu", "t", "we", "wi", "ll".
And i need add some postprocessing like this:
```python
# pip install fast-mosestokenizer
for seg in segments:
    text = seg["text"]
    word_timestamps = seg["word_timestamps"]
    
    tokens = self.tokenize(text)
    tokens = [re.match("[a-zA-Zа-яА-Я\\d]+", t) for t in tokens]
    tokens = [t.group(0) for t in tokens if t]
    
    ti = 0
    new_wt = []
    buff_wt = {}
    
    for wt in word_timestamps:
        word = re.match("[a-zA-Zа-яА-Я\\d]+", self.tokenize(wt['word'])[0])
    
        if word:
            wt['word'] = word.group(0)
    
            if not buff_wt:
                buff_wt = wt
            else:
                buff_wt["word"] = f"{buff_wt['word']}{wt['word']}"
    
            if buff_wt["word"] == tokens[ti]:
                ti += 1
                new_wt.append(buff_wt)
                buff_wt = {}
```

**Reply by jianfch** [reply]
> jumps from 48 -> 67

so the latter is the case. the original did overshoot the time

**Reply by eschmidbauer** [reply]
how did the word `around` jump to 67s ? Is that an issue with the [code](https://github.com/jianfch/stable-ts) or something else?

**Reply by jianfch** [reply]
> Additional. In my case on another language words more like a tokens. Something like this: "bu", "t", "we", "wi", "ll". And i need add some postprocessing like this:

yes, i forgot put in the notes that it actually when i say "word timestamps" i meant timestamp of each token
while most english tokens are individual words, compound words tend to get split into separate tokens. so i might have to add some postprocessing for that too.
thanks for the reminder!

**Reply by jianfch** [reply]
> how did the word `around` jump to 67s ? Is that an issue with the [code](https://github.com/jianfch/stable-ts) or something else?

that's just the model's prediction. if that is the last section of the audio, i suspect it has to do with how that section is being padded to 30seconds causing those bad predictions

**Reply by jianfch** [reply]
just fixed a major bug _that made stabilizing almost unusable_
also made it more robust to cases like the aforementioned 
![image](https://user-images.githubusercontent.com/28970749/193170936-6a278f79-be32-4d7f-848d-d3b492500670.png)
this is done by discarding any overshoots made at filtering logit stage _this should make it more robust in general too_
lastly added some quick handy functions, including one to generate a token level .srt directly from results
```python
from stable_whipser import results_to_srt

# after you get result from model
results_to_srt(result, 'audio.srt', word_level=True)
```

**Reply by jianfch** [reply]
what was the error you got? ts_logits is not necessary, its more for debugging purposes bc those are the probabilities of the timestamp tokens. ~~ill patched it make it disabled by default.~~

**Reply by ArtyomZemlyak** [reply]
Sorry, i just dont properly copy your code (I need, bc put it directly in whisper src) 
Fixed, and now works 
All good!

**Reply by eschmidbauer** [reply]
wow great improvement in latest changes. here is some example [output](https://gist.github.com/eschmidbauer/f7930e1b98992beff8681ab417504c38) 

I still see some timestamp overlaps though:
![image](https://user-images.githubusercontent.com/7139998/193286782-60dbf19a-277b-49cd-a12a-1d7662d432ab.png)


yet the unstable timestamps seem to have more accurate "selections" available
![image](https://user-images.githubusercontent.com/7139998/193287268-2603d2ab-45b3-4835-a7d4-629ef6852d7e.png)

Im using MFA  to compare `stable-ts` outputs to
![image](https://user-images.githubusercontent.com/7139998/193287388-9d4141bc-31b2-403c-907d-a6807fc86c17.png)

We aren't looking for precision - just need correct ordering. Wondering if you have any tips or ideas to prevent the overlapping timestamps? Really appreciate your work BTW!

**Reply by q00u** [reply]
>We aren't looking for precision

Well, _I_ am definitely looking for precision, because I want the timings to be consistent between transcripts.

**Reply by jianfch** [reply]
> We aren't looking for precision - just need correct ordering. Wondering if you have any tips or ideas to prevent the overlapping timestamps? Really appreciate your work BTW!

you can try the following:
```python
from stable_whisper import stabilize_timestamps

#with the modified model
# ts_num=7 is num of timestamps to get, so 7 is more than the default of 5
# stab=True  is disable stabilization so you can do it later with different settings
result = model.transcribe('audio.mp3', ts_num=7, stab=False) 
stab_segments = stabilize_timestamps(result ['segments'], aggressive=True ) # aggressive allows more variation
```
if this does reduces the overlapping, it might do it at the cost of accuracy and precision

but I'm working on another stabilization algorithm to put less emphasis on "top prediction" and use more of the unstable_timestamps.
On top of that, I'll test functionality to use results from multiple inferences with varying predictions to aggregate the best prediction.
_It's kind of like mixture of ensemble boosting and bagging style but for 1 model. It stems from how the same model can produce different but sometimes slightly more accurate predictions by introducing some noise to the input, but I think this benefits translation and possibly timestamps more than transcription because it's already near human-level for English_

**Reply by doesdev** [reply]
Sorry for adding noise here, but I just want to say thanks for the work you are doing on this Jian. It is greatly appreciated!

**Reply by jianfch** [reply]
Added new method for stabilizing token timestamps and made it the default.
old method -> new method
![image](https://user-images.githubusercontent.com/28970749/193392720-970e7ccb-8907-4754-b514-751e2920c7f2.png)

the new method does not adhere the top prediction, instead it looks at everything in `unstable_word_timestamps`
if you prefer the old method:
```python
stab_segments = stabilize_timestamps(result, top_focus=True)
```
or play around with the `ts_num` for `model.transcribe()` and the other params in `stabilize_timestamps()` to get other results

**Comment by q00u** [comment]
Came here to ask this, because the phrase-level timestamps are _wildly_ inaccurate, at least in the Medium model I used. I tried to transcribe a podcast with three speakers, each with their own discrete audio track.

So, three transcripts, which are then synced together into one script. This is the method I usually use, and it works quite well (in the other transcription service I use). You always know who is speaking, and it's transcribing from clean audio with only ever one speaker at a time.

But.

Currently, this is *unfeasible* with Whisper. There are frequent gaps of several seconds before or after the given phrase in a segment, making it impossible to know when any of the words in the segment were actually spoken, so syncing THREE cannot be done out of the box yet.

**Reply by Jxspa** [reply]
I've also found the timestamps to be extremely inaccurate (small/medium.en/medium model). They're regularly out by many seconds, and I've even had a few instances of them going backwards > `03:54.000 --> 03:34.720`.

**Comment by m-bain** [comment]
@q00u @Jxspa actually it can be done, see here https://github.com/m-bain/whisperX

**Comment by jongwook** [comment]
In addition to what others have introduced, I've made a demo of obtaining word-level timestamps using the cross-attention patterns in [the multilingual ASR notebook](https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb)

![image](https://user-images.githubusercontent.com/266841/210467619-4384e3d0-522b-4e84-8c64-60e6422eed26.png)

**Reply by ryanheise** [reply]
Oh, incidentally, dtw-python uses a license that, in terms of derivative works, is incompatible with almost all other licenses, including whisper's license (MIT). Maybe an alternative to dtw-python would be a better fit? There seem to be a few different dtw python libraries out there. Otherwise, at the very least, this notebook, which is stored under the same repository, would need its own separate license different from the license stored at the root directory of this repository.

**Reply by eschmidbauer** [reply]
should [the multilingual ASR notebook](https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb) work with large audio files? I'm getting a crash testing a file with audio length of `00:01:22.94`

```RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasLtMatmul( ltHandle, computeDesc.descriptor(), &alpha_val, mat1_ptr, Adesc.descriptor(), mat2_ptr, Bdesc.descriptor(), &beta_val, result_ptr, Cdesc.descriptor(), result_ptr, Cdesc.descriptor(), &heuristicResult.algo, workspace.data_ptr(), workspaceSize, at::cuda::getCurrentCUDAStream())```

**Reply by mu4farooqi** [reply]
Until Whisper merges that code to its main repository. I did it in my own branch. You can install and try.

https://github.com/mu4farooqi/whisper/tree/word_level_ts

**Reply by Jeronymous** [reply]
I improved the solution of @jongwook (thanks to him for the very useful notebook!) to be able to get cross-attention weights on the fly while the model is ran on segments of audio.
So there is no need to run the model twice (twice faster). It also solves memory issues that could occur on big signals, which @eschmidbauer reports.

Also I think that in the current version of the notebook by @jongwook there is an undesired shift of one token (the cross-attention weights computed on a given input token are relevant for the prediction of the next token).

Please find my code here: https://github.com/Jeronymous/whisper-timestamped

Any feedback welcome on this repo

**Reply by bfeist** [reply]
> I improved the solution of @jongwook (thanks to him for the very useful notebook!) to be able to get cross-attention weights on the fly while the model is ran on segments of audio. So there is no need to run the model twice (twice faster). It also solves memory issues that could occur on big signals, which @eschmidbauer reports.
> 
> Also I think that in the current version of the notebook by @jongwook there is an undesired shift of one token (the cross-attention weights computed on a given input token are relevant for the prediction of the next token).
> 
> Please find my code here: https://github.com/Jeronymous/whisper-timestamped
> 
> Any feedback welcome on this repo

Thanks very much for this. Could you possibly provide some notes on how to execute your repo from the command line?

**Reply by Jeronymous** [reply]
Sure. I completed the README. It should be clearer now in github repo.
Basically, there is a CLI "whisper_timestamped" that works similarly as the "whisper" CLI (it will produce more things, and does not support some options related to beam search)

**Comment by Gigio2k** [comment]
i'm still get an error : 

$ whisper_timestamped videoplayback.mp3 --model tiny --language it --accurate --verbose True

(audio length is around 3 minutes) 

...
[02:09.500 --> 02:10.340] Tolfika,
[02:10.720 --> 02:11.040] perdere.
Traceback (most recent call last):
  File "/home/ubuntu/.local/bin/whisper_timestamped", line 8, in <module>
    sys.exit(cli())
  File "/home/ubuntu/.local/lib/python3.10/site-packages/whisper_timestamped/transcribe.py", line 2156, in cli
    result = transcribe_timestamped(
  File "/home/ubuntu/.local/lib/python3.10/site-packages/whisper_timestamped/transcribe.py", line 253, in transcribe_timestamped
    (transcription, words) = _transcribe_timestamped_naive(model, audio,
  File "/home/ubuntu/.local/lib/python3.10/site-packages/whisper_timestamped/transcribe.py", line 1165, in _transcribe_timestamped_naive
    assert len(segment_tokens_check) < len(segment["tokens"]) and segment_tokens_check[:-1] == segment["tokens"][:len(segment_tokens_check)-1], \
AssertionError: Got inconsistent tokens:  != �a, joJo, Bachelor.
